pkg/ovs/openflow/ofctrl_action.go

```diff
@@ -286,6 +287,93 @@ func (a *ofFlowAction) Group(id GroupIDType) FlowBuilder {
 	return a.builder
 }
 
+//  Learn is an action which adds or modifies a flow in an OpenFlow table.
+func (a *ofFlowAction) Learn(id TableIDType, priority uint16, idleTimeout, hardTimeout uint16, cookieID uint64) LearnAction {
+	la := &ofLearnAction{
+		flowBuilder: a.builder,
+		nxLearn:     ofctrl.NewLearnAction(uint8(id), priority, idleTimeout, hardTimeout, 0, 0, cookieID),
+	}
+	return la
+}
+
+// ofLearnAction is used to describe actions in the learn flow.
+type ofLearnAction struct {
+	flowBuilder *ofFlowBuilder
+	nxLearn     *ofctrl.FlowLearn
+}
+
+func (a *ofLearnAction) DeleteLearned() LearnAction {
+	a.nxLearn.DeleteLearnedFlowsAfterDeletion()
+	return a
+}
+
+// MatchLearnedTCPDstPort specifies that the tcp_dst field in the learned flow must match the tcp_dst of the packet currently being processed.
+func (a *ofLearnAction) MatchLearnedTCPDstPort() LearnAction {
+	ethTypeVal := make([]byte, 2)
+	binary.BigEndian.PutUint16(ethTypeVal, 0x800)
+	ipTypeVal := make([]byte, 2)
+	ipTypeVal[1] = byte(ofctrl.IP_PROTO_TCP)
+	a.nxLearn.AddMatch(&ofctrl.LearnField{Name: "NXM_OF_ETH_TYPE"}, 2*8, nil, ethTypeVal)
+	a.nxLearn.AddMatch(&ofctrl.LearnField{Name: "NXM_OF_IP_PROTO"}, 1*8, nil, ipTypeVal)
+	a.nxLearn.AddMatch(&ofctrl.LearnField{Name: "NXM_OF_TCP_DST"}, 2*8, &ofctrl.LearnField{Name: "NXM_OF_TCP_DST"}, nil)
+	return a
+}
+
+// MatchLearnedUDPDstPort specifies that the udp_dst field in the learned flow must match the udp_dst of the packet currently being processed.
+func (a *ofLearnAction) MatchLearnedUDPDstPort() LearnAction {
```

Comment:
Can MatchLearnedTCPDstPort and MatchLearnedUDPDstPort share some code?

Reply:
Updated, PTAL.